### ISSUE TITLE
Add atob, btoa, structuredClone, queueMicrotask and performance

### ISSUE THREAD

**Issue by bnoordhuis** [quickjs-ng/quickjs#16]
Continuing from https://github.com/saghul/txiki.js/issues/418:

To avoid duplication across embedders, implement atob, btoa, structuredClone<sup>1</sup>, queueMicrotask and performance as quickjs-ng C APIs and/or JS globals. We also have to add DOMException but that seems okay to me. We can make them opt-in when the JSContext is created.
 
<sup>1</sup> TBD how to handle transferables

- [ ]   globalThis.[atob()](https://html.spec.whatwg.org/multipage/webappapis.html#dom-atob)
- [ ]   globalThis.[btoa()](https://html.spec.whatwg.org/multipage/webappapis.html#dom-btoa)
- [x]   globalThis.[performance](https://w3c.github.io/hr-time/#dom-windoworworkerglobalscope-performance).[now()](https://w3c.github.io/hr-time/#dom-performance-now)
- [x]   globalThis.[performance](https://w3c.github.io/hr-time/#dom-windoworworkerglobalscope-performance).[timeOrigin](https://w3c.github.io/hr-time/#dom-performance-timeorigin)
- [x]   globalThis.[queueMicrotask()](https://html.spec.whatwg.org/multipage/timers-and-user-prompts.html#dom-queuemicrotask)
- [ ]   globalThis.[structuredClone()](https://html.spec.whatwg.org/multipage/structured-data.html#dom-structuredclone)

**Comment by saghul** [comment]
Doesn't the performance.now put us into a position where we need to start branching out to platform specific code?

Is there a minimal implementation we could support and then decide to use it at context creation time with flags perhaps?

**Comment by M0n7y5** [comment]
Perhaps you can fetch counter from processor? Maybe RDTSC on X86 would help? That would be compiler specific which is not that bad imho.

**Comment by M0n7y5** [comment]
Nwm it needs TSC which is platform specific. That means it belongs to std runtime implementation.

**Comment by bnoordhuis** [comment]
Node.js uses uv_hrtime() and that's basically a wrapper around clock_gettime(CLOCK_MONOTONIC) or QueryPerformanceCounter(). Seems okay to me to use that here as well.

**Comment by saghul** [comment]
Yep I had that in mind too. Does also work on macOS and BSDs? I haven't checked the implementation in a while :-) 

If it's simple let's do it !

**Comment by bnoordhuis** [comment]
Even macOS has clock_gettime as of 10.12 so I don't think there are any blockers.

**Comment by saghul** [comment]
Awesome!

**Comment by M0n7y5** [comment]
Yeah but clock_gettime is not everywhere just like QueryPerformanceCounter(). I suggest to not make performance or any other platform specific things to be part of the core library. Sort of extension to standard lib is more useful. With proper defines which allows you to opt-out of these things if needed.

**Comment by saghul** [comment]
Why not? Implementing `Atomics` already requires some if-defing around, and if all we need is to support 2 implementations of `performance.now`, that's a drop in a bucket compared to the size of quickjs.c.

**Comment by bnoordhuis** [comment]
> clock_gettime is not everywhere

Where is it missing? Linux, macOS, (Free|Net|Open|Dragonfly)BSD, Solaris, Fuchsia, QNX and Emscripten all have it.

**Comment by saghul** [comment]
Brain dump, since I've been thinking about these a bit the past week.

Both atob/btoa and structuredClone depend on DOMException for error reporting, and if we are to do it we might as well make it compliant, so we should have DOMException I guess.

I noticed no other engine provides structuredClone, it seems to always be implemented by the embedder for some reason. If we can do it, then why not? (IMHO).

Last, atob/btoa require a base64 encoder/decoder, which we will also need to add for https://github.com/tc39/proposal-arraybuffer-base64 so I guess it ties it up nicely...

**Comment by graham** [comment]
I took a shot at implementing atob/btoa here: https://gist.github.com/graham/0f6f0f7de6c2839155c7f36ee7625588

It uses come c++ stdlib and an external library for printing, but I'll remove all of that before a create a pr.

I'm a little confused about how I should go about testing this? I could write something in `api-test.c`, or I could simply add the atob/btoa to the globalThis and write a javascript test in `/tests/`

I don't know enough about how the arraybuffer work is going to know if this is helpful or a distraction.

__Actual Question: Assuming I'm going to make a pr to add atob and btoa; where would you like me to implement the tests to ensure it works correctly?__

Alternatively, if this is already in the works I can find another issue to focus on.

**Comment by saghul** [comment]
Looks good!

I'd add them to the global object by default, perhaps with the option to not do that with some helper JS_AddIntrinsicAtob for those who want to use the raw context.

This way tests can be written in JS.

Error handling is something we need to solve though. In the spec they can throw DOMException, so we likely need to implement that first or as part of this work.

**Comment by graham** [comment]
I took another shot at this: https://github.com/quickjs-ng/quickjs/pull/1032

Removed all the c++ and tried to make it as much like the other modules as I could. My initial implementation is pretty basic, after completing it I found: https://stackoverflow.com/questions/342409/ as well as a couple other articles about how to make this more performant if there is interest.

This is my first contribution so feel free to be extra picky if I missed a style or organization detail. (any feedback is welcome)

**Comment by sergeevabc** [comment]
Guys, I have QuickJS version 2025-04-26 and try to execute it on Windows 7 as follows
``` console
$ qjs script.txt
ReferenceError: 'atob' is not defined
    at <eval> (script.txt:9:13)
```

However it works with `node` and `deno` instead of `qjs`, and I see the output.

Does it mean QuickJS is missing `atob()` function? :(

**Comment by bnoordhuis** [comment]
Yes, but see #1143.

**Comment by sergeevabc** [comment]
@bnoordhuis, thank you for pointing out the encouraging work of @aabbdev
I hope it will work on Windows 7.

**Comment by aabbdev** [comment]
@sergeevabc When I have time; to be merged I need to pull the master the implementation is ready
